File: client/src/api/responsibilities.ts
import { apiFetch } from "./client";

export interface Responsibility {
  id: number;
  name: string;
  actor_label: string | null;
  difficulty_variant: string | null;
  requires_role: string | null;
  requires_prior_experience: boolean;
  description: string | null;
  confidence: string;
  note_line: string | null;
  last_updated: string | null;
}

export type ResponsibilityInput = Omit<Responsibility, "id" | "last_updated">;

export function getResponsibilities(): Promise<Responsibility[]> {
  return apiFetch<Responsibility[]>("/api/responsibilities");
}

export function getResponsibility(id: number): Promise<Responsibility> {
  return apiFetch<Responsibility>(`/api/responsibilities/${id}`);
}

export function createResponsibility(data: ResponsibilityInput): Promise<Responsibility> {
  return apiFetch<Responsibility>("/api/responsibilities", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function updateResponsibility(
  id: number,
  data: Partial<ResponsibilityInput>,
): Promise<Responsibility> {
  return apiFetch<Responsibility>(`/api/responsibilities/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export function deleteResponsibility(id: number): Promise<void> {
  return apiFetch<void>(`/api/responsibilities/${id}`, { method: "DELETE" });
}

File: server/routes/responsibilities.py
from flask import Blueprint, jsonify, request

from extensions import db
from models import Responsibility

responsibilities_bp = Blueprint(
    "responsibilities", __name__, url_prefix="/api/responsibilities"
)

REQUIRED_FIELDS = ["name"]
UPDATABLE_FIELDS = [
    "name",
    "actor_label",
    "difficulty_variant",
    "requires_role",
    "requires_prior_experience",
    "description",
    "confidence",
    "note_line",
    "type",
]


@responsibilities_bp.get("")
def list_responsibilities():
    return jsonify([r.to_dict() for r in Responsibility.query.all()])


@responsibilities_bp.get("/<int:responsibility_id>")
def get_responsibility(responsibility_id):
    responsibility = db.session.get(Responsibility, responsibility_id)
    if responsibility is None:
        return jsonify(error="Responsibility not found"), 404
    return jsonify(responsibility.to_dict())


@responsibilities_bp.post("")
def create_responsibility():
    data = request.get_json(silent=True) or {}
    missing = [f for f in REQUIRED_FIELDS if not data.get(f)]
    if missing:
        return jsonify(error=f"Missing fields: {', '.join(missing)}"), 400

    # Solo pasamos al constructor los campos que realmente llegaron en el
    # body, para no pisar los defaults del modelo (ej. confidence) con None
    fields = {f: data[f] for f in UPDATABLE_FIELDS if f in data}
    responsibility = Responsibility(**fields)
    db.session.add(responsibility)
    db.session.commit()
    return jsonify(responsibility.to_dict()), 201


@responsibilities_bp.put("/<int:responsibility_id>")
def update_responsibility(responsibility_id):
    responsibility = db.session.get(Responsibility, responsibility_id)
    if responsibility is None:
        return jsonify(error="Responsibility not found"), 404

    data = request.get_json(silent=True) or {}
    for field in UPDATABLE_FIELDS:
        if field in data:
            setattr(responsibility, field, data[field])
    db.session.commit()
    return jsonify(responsibility.to_dict())


@responsibilities_bp.delete("/<int:responsibility_id>")
def delete_responsibility(responsibility_id):
    responsibility = db.session.get(Responsibility, responsibility_id)
    if responsibility is None:
        return jsonify(error="Responsibility not found"), 404

    db.session.delete(responsibility)
    db.session.commit()
    return "", 204
